src/katgpucbf/xbgpu/engine.py

```diff
@@ -282,23 +283,30 @@ def __init__(
         super().__init__(katcp_host, katcp_port)
         self._cancel_tasks: list[asyncio.Task] = []  # Tasks that need to be cancelled on shutdown
 
+        # B-engine doesn't work yet
+        assert len(outputs) == 1
+        assert isinstance(outputs[0], XOutput)
+
         if sample_bits != 8:
             raise ValueError("sample_bits must equal 8 - no other values supported at the moment.")
 
-        if channel_offset_value % n_channels_per_stream != 0:
-            raise ValueError("channel_offset must be an integer multiple of n_channels_per_stream")
+        for output in outputs:
+            if channel_offset_value % output.channels_per_substream != 0:
+                raise ValueError(f"{output.name}: channel_offset must be an integer multiple of channels_per_substream")
+            if output.channels_per_substream != n_channels_per_stream:
+                raise ValueError(
+                    f"{output.name} has {output.channels_per_substream} channels-per-substream, \
+                                 which is not compatible with the global value ({n_channels_per_stream})"
+                )
+            # TODO: Do we need to ensure all output.dst addresses do not overlap?
```

Comment:
I wouldn't bother. If the user wants it, they can have it. In production the addresses are generated by the computer, so you don't need to protect the user from finger trouble.